src/pybind/mgr/dashboard_v2/services/ceph_service.py

```diff
@@ -1,32 +1,112 @@
+
+            try:
+                status = json.loads(status['json'])
+            except ValueError:
+                status = {}
+
+            instance_id = metadata['instance_id']
+            if id == instance_id:
+                # new version that supports per-cluster leader elections
+                id = metadata['id']
+
+            # extract per-daemon service data and health
+            daemon = {
+                'id': id,
+                'instance_id': instance_id,
+                'version': metadata['ceph_version'],
+                'server_hostname': hostname,
+                'service': service,
+                'server': server,
+                'metadata': metadata,
+                'status': status
+            }
+            daemon = dict(daemon, **_get_daemon_health(daemon))
+            daemons.append(daemon)
+
+    return sorted(daemons, key=lambda k: k['instance_id'])
+
+
+def _get_daemon_health(daemon):
+    health = {
+        'health_color': 'info',
+        'health': 'Unknown'
+    }
+    for _, pool_data in daemon['status'].items():  # TODO: simplify
+        if (health['health'] != 'error' and
+            [k for k, v in pool_data.get('callouts', {}).items()
+             if v['level'] == 'error']):
+            health = {
+                'health_color': 'error',
+                'health': 'Error'
+            }
+        elif (health['health'] != 'error' and
+              [k for k, v in pool_data.get('callouts', {}).items()
+               if v['level'] == 'warning']):
+            health = {
+                'health_color': 'warning',
+                'health': 'Warning'
+            }
+        elif health['health_color'] == 'info':
+            health = {
+                'health_color': 'success',
+                'health': 'OK'
+            }
+    return health
+
+
+class CephServiceMixin(object):
 
-class CephService(Service):
     @classmethod
     def get_service_map(cls, service_name):
-        service_map = {}
-        for server in cls.mgr.list_servers():
-            for service in server['services']:
-                if service['type'] == service_name:
-                    if server['hostname'] not in service_map:
-                        service_map[server['hostname']] = {
-                            'server': server,
-                            'services': []
-                        }
-                    inst_id = service['id']
-                    metadata = cls.mgr.get_metadata(service_name, inst_id)
-                    status = cls.mgr.get_daemon_status(service_name, inst_id)
-                    service_map[server['hostname']]['services'].append({
-                        'id': inst_id,
-                        'type': service_name,
-                        'hostname': server['hostname'],
-                        'metadata': metadata,
-                        'status': status
-                    })
-        return service_map
+        return _get_service_map(cls.mgr, service_name)
```

Comment:
I don't see the benefit of moving this code outside the `CephServiceMixin` class. I don't see a problem of calling other mixins functions inside a mixin.

Reply:
> I don't see a problem of calling other mixins functions inside a mixin.

There's a problem with the `mgr` instance, which is not yet there when calling `CephServiceMixin.<method>` from `CephRbdMirrorMixin`, for instance. Do you have a better solution for this, instead of outsourcing the methods to the module like I did?

Reply:
Here's a good case where mixins fail to provide the necessary infrastructure. The idea of these "services" classes is to implement the "business" logic common to several controllers. It should be natural for services to call each other in a compositional way. I should be able to create a more complex service by composition of functions of existing services.

With the current mixin solution, services logic/implementation is not independent of controllers because they require their `mgr` instance. To solve your particular problem, I would do this:

```
CephServiceMixin.mgr = cls.mgr
return CephServiceMixin.get_service_map(service_name)
```

But it is still ugly though :(

Having functions declared global in the module scope, will fail to create an organized structure of the code, unless we add big comment notes separating each set of functions according to the mixin class they primary belong.

Reply:
This is the same problem we have with controllers, the `mgr` instance is injected to each controller (but it could be passed in the constructor!). Maybe we could do the same for the mixins. Would feel better than injecting the attribute right before the call.